### DISCUSSION TITLE
Raspberry Pi boards suitable for use with Basecamp

### DISCUSSION THREAD

**Discussion by open-stemware** [cfs-tools/cfs-basecamp#89] [General]
This discussion originated in Issue #88. This issue was closed and has been reproduced here to start a discussion.  

**WIQu**
Which Raspberry Pi boards suitable for use with Basecamp?

**Open STEMware:**
I have used Basecamp on RPI 3B,3B+, 4B, and 5. Note if you are using the most recent RPI OS called Bookworm then you may have to use Python's virtual environment. I heard of pip package installation issues if you don't.

PI_IOLIB is designed for RPIs 4B and earlier and you select your model in the config.h file. PI_IOLIB is someone else's library that I wrapped to work as a cFS library. It serves as a good example for C interfacing in the GPIO_DEMO app but I have not used many of its features.

Many users do not want to write C interfaces, so I am working on new libs/apps that will allow simple python scripts to interface with hardware and communicate with a basecamp app using JSON messages. Initial versions of these libs/apps will be released in July.

**WIQu**
Thank you for your answer！If I want to port cFS to the TI hardware processor platform, what kind of learning and work would be required? Could you provide some suggestions？

**Comment by open-stemware** [comment]
I answered based on my own Raspberry Pi experience with Basecamp.  As far as I know the cFS has been run on all versions of the Pi.  

I do not have experience with the cFS on TI processors. [cFS Discussions](https://github.com/nasa/cFS/discussions) is a good place to post porting questions.

**Comment by alanc-gsfc** [comment]
We do need a cFS porting guide to help answer these types of questions. Some basics:
- Operating System: cFS supports Linux/POSIX, VxWorks, and RTEMS - I would want to start with a board that has one of these operating systems running on it.
- Resources: A cFS system runs on a 32 or 64 bit processor. I don't think 16 bit processors are supported. The amount of memory depends on the apps and your configuration but on a Real Time OS (VxWorks or RTEMS), you should have at least 8 megabytes of RAM. Linux is going to need more RAM. This will leave out MCU class processors that usually have less than 1 Megabyte of internal RAM.

Addressing TI processors: What CPU and Architecture is it? I have run cFS on a BeagleBone black on both RTEMS and Linux.

**Reply by 1763545734** [reply]
Thank you very much for your reply.

I will be using the ARM® Cortex™-R4F 32-bit RISC CPU and architecture. Could you please provide the method for running cFS on a BeagleBone Black? As I am new to cFS, this might take a considerable amount of time. I will share the issues I encounter and the solutions I find during the porting process.

Thank you again for your help!

**Reply by alanc-gsfc** [reply]
Instructions for running on the BeagleBone Black vary depending on the operating system. Since you are new to cFS, I would recommend starting with the default Linux OS they provide, then build and run just as you would on Linux.

To work your way up to a fully embedded system, you could do the following:
1. Using the BeagleBone black as a desktop (or remote) ssh Linux system, clone, build and run cFS just as you would on a Linux desktop/VM.
2. Use the cross compilation toolchain for the beagle bone black on your development Linux desktop/vm to build the cFS for the Beaglebone Linux. You will need a ARM cross compiler for this. The toolchain is called :toolchain-arm-cortexa8_neon-linux-gnueabi.cmake
Once you cross compile cFS, you can transfer build/exe/cpu1 over to the beagle bone Linux system and hopefully it will run.

If you plan on using Linux ( sounds like no with the Cortex-R4F), then you don't have to start using a RTOS.

But I think you may need a RTOS on the R4F.

How much memory does the system with the Cortex R4F have?? You should have at least 4 Megabytes of RAM, hopefully 8 to 16 Megabytes. The cFS will not run in a microcontroller with only 256 or 512Kbytes of RAM.
